
# This script separates the alignments by protein domains and calculates nucleotide diversity (pi) for each domain.
# The structure-domain information is taken from Interpro website (See Methods in the manuscript).
# To avoid redundant information, the alignments by domain are not included in the repository. 
# They can be generated by splitting the alignments (output of script 08) using structure-domain information.


library("ape")
library("pegas")

path <- "Local/path/to/alignments/by/domains"
path_out <- "Data/09.Nucleotide_Diversity/out/"
nona <- c("OT", "VT") # Nonapeptides
dom_nona <- c("signal_peptide", "nonapeptide", "neurophysin", "copeptin") # Domains of nonapeptides
recep <- c("OTRa", "OTRb.tr2", "VTR1Aa.tr1", "VTR1Ab", "VTR2Aa", "VTR2Ab.tr2", "VTR2Ba", "VTR2Bb.tr1") # Receptors
# Only Receptors with the canonical GPCR stricture are used
dom_recept <- c("nter", "tm1", "il1", "tm2", "el1", "tm3", "il2", "tm4", "el2", "tm5", "il3", "tm6", "el3", "tm7", "cter") 
# Domains of receptors


# Nonapeptides
df_out <- data.frame(matrix(ncol = length(dom_nona) + 1))
colnames(df_out) <- c("gene", dom_nona)

for (gen in nona) {

  vec <- c(gen)
  
  for (dom in dom_nona) {
    
    # Read alignment
    dna <- ape::read.dna(paste(path, gen, "/", dom, ".fa", sep = ""), format = "fasta")
    # Calculate nucleotide diversity
    nd <- nuc.div(dna, variance = F, pairwise.deletion = T)
    
    vec <- c(vec, nd)
    
  }
  
  df_out <- rbind(df_out, vec)
    
}

df_out <- na.omit(df_out)

write.table(df_out, file = paste0(path_out, "pi_nonapeptides_domains.txt"), sep = "\t", col.names = T, quote = F, row.names = F)



# Receptors

df_out <- data.frame(matrix(ncol = length(dom_recept) + 1))
colnames(df_out) <- c("gene", dom_recept)

for (gen in recep) {
  
  vec <- c(gen)
  
  for (dom in dom_recept) {
    
    # Read alignment
    dna <- ape::read.dna(paste(path, gen, "/", dom, ".fa", sep = ""), format = "fasta")
    # Calculate nucleotide diversity
    nd <- nuc.div(dna, variance = F, pairwise.deletion = T)
    
    vec <- c(vec, nd)
    
  }
  
  df_out <- rbind(df_out, vec)
  
}

df_out <- na.omit(df_out)

write.table(df_out, file = paste0(path_out, "pi_receptors_domains.txt"), sep = "\t", col.names = T, quote = F, row.names = F)
