#!/bin/bash
# This scripts performs positive selection on the aligned sequences in script 08, using Hyphy FEL method.
# It uses the gene trees generated in script 10 and the species tree from Ronco et al. (2021).
# Hyphy version 2.5


source Scripts/functions_bash.sh # To access custom functions defined in functions_bash.sh

# Paths
path_aln="Data/08.Alignments/out/"
path_pos_sel="Data/12.FEL_pos_selection/out/"
path_trees="Data/10.Gene_Trees/out/trees/"
path_gene_dNdS="Data/11.dS_ratios/out/gene-wide_dNdS/"
# Files
spp_tree="Data/12.FEL_pos_selection/in/spp_tree.txt" # Taken from Ronco et al. (2021)
# Array
genes=( $(cat "Data/arrays/transcripts.txt") )


cd "${path_pos_sel}"

# Site-specific positive selection
for gen in "${genes[@]}" ; do

    # Define alignment file and gene tree file
    aln_file="${path_aln}/${gen}/aln_${gen}_CDS_p2n.fa"
    gene_tree="${path_trees}/${gen}/${gen}_cds_tree/aln_${gen}_CDS_p2n.fa.treefile"

    # Create a directory for each gene
    mkdir "${gen}" ; cd "${gen}"

    # Remove Reference genome sequence from the alignment
    single_line_fa "${aln_file}"
    sed '/Orenil/{N;d;}' "${path_aln}/${gen}/aln_${gen}_CDS_p2n.fa" > ./inp/aln_"${gen}".fa
    # The alignments are not shown to avoid redundant information. They are the same sequences as in folder 08, excluding Orenil sequence

    #FEL positive selection
    hyphy fel --alignment ./inp/aln_"${gen}".fa --tree "${gene_tree}" --pvalue 0.05 ENV=TOLERATE_NUMERICAL_ERRORS=1 --output ./out/site_FEL/json_files/"${gen}"_FEL_gene.json # gene_tree
    hyphy fel --alignment ./inp/aln_"${gen}".fa --tree "${spp_tree}" --pvalue 0.05 ENV=TOLERATE_NUMERICAL_ERRORS=1 --output ./out/site_FEL/json_files/"${gen}"_FEL_phyl.json # phyl_tree

    #Parse output
    python Scripts/functions_python.py hyphy_parse_fel ./out/site_FEL/json_files/"${gen}"_FEL_gene.json ./out/gene_FEL "${gen}"
    python Scripts/functions_python.py hyphy_parse_fel ./out/site_FEL/json_files/"${gen}"_FEL_phyl.json ./out/phyl_FEL "${gen}"
    # After parsing, the information is summarized in CSV files found in folder out/site_FEL/.
    # An overview of the site-specific positive selection results, together with allele and domain information, can be found in out/overviews_positive_selection/

    cd "${path_pos_sel}"

done


# Gene-wide dNdS. It is taken from JSON file generated by Hyphy FEL
models=( phyl gene )

for mod in "${models[@]}" ; do
    
    # Extract dNdS
    for gen in "${genes[@]}" ; do
    
        # Define JSON file
        json_file="out/site_FEL/json_files/${gen}_FEL_${mod}.json"
        
        # Specific function to extract dNdS value and number of sites in FEL JSON output file
        dNdS=$( cat "${json_file}" | grep -A32 "non-synonymous/synonymous" | grep -A1 "synonymous" | grep -v ":" | sed "s/\[//g" | cut -d"," -f1 )
        sites=$( cat "${json_file}" | grep -A32 "non-synonymous/synonymous" | grep "sites" | sed "s/,//g" | cut -d":" -f2 )
        
        # Echo to output file
        echo -e "${gen}\t${dNdS}\t${sites}"
        
    done > "${path_gene_dNdS}/dNdS_${mod}.txt"
    
done

exit
